core/src/main/java/org/apache/iceberg/variants/ShreddedObject.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.variants;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.Map;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.Pair;
+import org.apache.iceberg.util.SortedMerge;
+
+/**
+ * A variant Object that handles full or partial shredding.
+ *
+ * <p>Metadata stored for an object must be the same regardless of whether the object is shredded.
+ * This class assumes that the metadata from the unshredded object can be used for the shredded
+ * fields. This also does not allow updating or replacing the metadata for the unshredded object,
+ * which could require recursively rewriting field IDs.
+ */
+class ShreddedObject implements VariantObject {
+  private final SerializedMetadata metadata;
+  private final SerializedObject unshredded;
+  private final Map<String, VariantValue> shreddedFields = Maps.newHashMap();
+  private SerializationState serializationState = null;
+
+  ShreddedObject(SerializedMetadata metadata) {
+    this.metadata = metadata;
+    this.unshredded = null;
+  }
+
+  ShreddedObject(SerializedObject unshredded) {
+    this.metadata = unshredded.metadata();
+    this.unshredded = unshredded;
+  }
+
+  public void put(String field, VariantValue value) {
+    Preconditions.checkArgument(
+        metadata.id(field) >= 0, "Cannot find field name in metadata: %s", field);
+
+    // allow setting fields that are contained in unshredded. this avoids read-time failures and
+    // simplifies replacing field values.
+    shreddedFields.put(field, value);
+    this.serializationState = null;
+  }
+
+  @Override
+  public VariantValue get(String field) {
```

Comment:
I see what we are implementing in ShreddedObject: basically we are providing the same interface get(String field) as regular VariantObject.

Given the following example, assume `event.location.latitude` is shredded while `event.location.longitude` is not. How do we model shredded `event` object - where to place the field `location` ? 




> event {
>  event_id;
>   location {
>      latitude;
>      longitude;
>   }
> }

Reply:
If `latitude` is shredded and `longitude` is not, then `location` must be a partially shredded object. That object would contain `longitude` in the `value` field and would have a `typed_value` group that contains a `latitude` group that has a `value` and `typed_value` pair.

And because `location` is a partially shredded object, `event` is also a partially shredded object that has a shredded `location` group.